server/aesdsocket.c

```diff
@@ -23,217 +23,313 @@
 #include <syslog.h>
 #include <net/if.h>
 #include <sys/stat.h>
+#include <string.h>
+#include <sys/queue.h>
+#include <pthread.h>
+#include <sys/time.h>
 
 
 #define DOMAIN AF_INET
 #define TYPE SOCK_STREAM
 #define PROTOCOL 0
 #define PORT 9000
-#define BUFFER_SIZE 1024
+//#define BUFFER_SIZE 1024
 
 
 int file_fd;
 int sockfd;
+pthread_mutex_t lock;
+
+typedef struct node
+{
+    pthread_t thread;
+    int clientfd;
+    bool thread_complete_status;
+    char client_ip[INET6_ADDRSTRLEN];
+    TAILQ_ENTRY(node) entries;
+}thread_data_t;
+
+typedef TAILQ_HEAD(head_s, node) head_t;
+head_t head;
+
+/* appended timestamp to the /var/tmp/aesdsocketdata file
+*  every 10s using RFC 2822 complaint strftime format,
+*  followed by a newline
+*/
+void time_stamp()
+{
+    printf("starting timestamp\n");
+    time_t ct;
+    char time_buffer[100];
+    time(&ct);
+    //RFC 2822 complaint strftime format,followed by a newline.
+    strftime(time_buffer, sizeof(time_buffer), "timestamp: %Y-%m-%d %H:%M:%S\r\n", localtime(&ct));
+    //char *timestamp = "timestamp: ";
+    // appending timestamp to the file
+    //strcat(timestamp, time_buffer);
+    printf("%s", time_buffer);
+    lseek(file_fd, 0, SEEK_END);
+    int count_bytes = strlen(time_buffer);
+    //writing to the file
+    int write_bytes_count = write(file_fd, time_buffer, count_bytes);
+    if (write_bytes_count != count_bytes){
+        printf("write unsuccessful\n");
+        //perror("write unsuccessful\n");
+    }
+    printf("exiting timestamp\n");
+}
+
+// free memory
+void free_memory()
+{
+    thread_data_t *temp;
+    while(!TAILQ_EMPTY(&head))
+    {
+        temp = TAILQ_FIRST(&head);
+        TAILQ_REMOVE(&head, temp, entries);
+        free(temp);
+    }
+}
 
 void handle_signal(int signum)
 {
-    printf("Caught signal %d", signum);
+    printf("Caught signal %d\n", signum);
+    if (signum == SIGALRM)
+    {
+        printf("caught SIGALRM\n");
+        time_stamp();
+        //unsigned int alarm (unsigned int seconds);
+        alarm(10);
+    }
     if ((signum == SIGINT) || (signum == SIGTERM)) {
+        printf("Caught signal %d\n", signum);
         close(file_fd);
         close(sockfd);
         remove("/var/tmp/aesdsocketdata");
-        exit(0);
+        free_memory();
+        exit (0);
```

Comment:
@TejaswiniLakshminarayan exiting here will result in a memory leak under some scenarios.  See my other comments about this.